HPCC-24546 Support for Hashicorp vaults and kubernetes secrets

afishbeck opened the pull request:
Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

afishbeckln committed:
HPCC-24546 Support for Hashicorp vaults and kubernetes secrets

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: 1a0059bea094181a68073568730dafed81f7b046
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-30.ca-central-1.compute.internal

afishbeck commented:
@richardkchapman @ghalliday please review.

The helm/examples/secrets/README.md file is meant to be a very easy to use walk through of setting up and using vault and kubernetes secrets.. please include usability comments in your review.

I've kept the secret and vault functions in jutil.cpp for now, but can move them once we've agreed on the approach.

I've added open source cpp-http code for my REST calls.  The nice thing is it's all in one header file.  But when I move the code I could use our CHttpClient code.  I didn't want to link esphttp to jlib.  And this library is a bit more low level.

afishbeck commented:
Another option for REST calls could be boost.. I didn't want to deal with large dependencies though.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-30.ca-central-1.compute.internal
Sha: 1a0059bea094181a68073568730dafed81f7b046
Number of Error(s): 4
/home/centos/smoketest/PR-14122/HPCC-Platform/system/httplib/httplib.h:211:2: error: #error Sorry, OpenSSL versions prior to 1.1.1 are not supported
checking for archiver @FILE support... make[2]: *** [system/jlib/CMakeFiles/jlib.dir/build.make:635: system/jlib/CMakeFiles/jlib.dir/jutil.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1302: system/jlib/CMakeFiles/jlib.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2020-08-28_16-29-04
time logs started
Prep time   :     9 sec (00:00:09)
Build time  :   448 sec (00:07:28)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   457 sec (00:07:37)
time logs end

[Build error:](https://gist.github.com/90b8606c7806cf378c20cbd8945e34a0)
Build log [Build log](https://gist.github.com/HPCCSmoketest/90b8606c7806cf378c20cbd8945e34a0#file-result-20-08-28-16-21-13-log)

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: b7ccf82891d1ba5f01cfe92d4857e0fdbdea2506
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-96.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-96.ca-central-1.compute.internal
Sha: b7ccf82891d1ba5f01cfe92d4857e0fdbdea2506
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |144 |0 |1 |0 |76 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Unittest error(s):
```
Run:  1   Failures: 0   Errors: 1
1) test: JlibSecretTest::test (E) 
uncaught exception of unknown type
ConfigMgr 2.0 - Load and Parse - Verifying configuration XSDs are compliant...
*** ConfigMgr 2.0 - Templates - Loading Only ***
*** ConfigMgr 2.0 - Templates - Inputs ***
*** ConfigMgr 2.0 - Templates - Variable substitution tests ***
*** ConfigMgr 2.0 - Templates - Execute Templates ***
*** ConfigMgr 2.0 - Templates - Find Node Tests ***
*** ConfigMgr 2.0 - Templates - Create Node Tests ***

```

Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |30 sec  (00:00:30) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |25 sec  (00:00:25) |
|test (hthor) |929 |929 |0 |345 sec  (00:05:45) |
|test (thor) |842 |842 |0 |756 sec  (00:12:36) |
|test (roxie) |1008 |1008 |0 |402 sec  (00:06:42) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  793 sec (00:13:13) |  138 sec (00:02:18) |   23 sec (00:00:23) |   21 sec (00:00:21) | 1863 sec (00:31:03) |   18 sec (00:00:18) | 2866 sec (00:47:46) |

afishbeck commented:
Removed tests that were only meant for my environment, and updated what seems to be arbitrary cpp-httplib openssl version deprecation.

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: 718e763fc11752ae2eebaec00a647fa7387b8595
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-205.ca-central-1.compute.internal

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Did not build yet.
Commit ID: 718e763fc11752ae2eebaec00a647fa7387b8595
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-205.ca-central-1.compute.internal
Sha: 718e763fc11752ae2eebaec00a647fa7387b8595
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |72 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |41 sec  (00:00:41) |
|setup (roxie) |11 |11 |0 |16 sec  (00:00:16) |
|test (hthor) |929 |929 |0 |336 sec  (00:05:36) |
|test (thor) |842 |842 |0 |747 sec  (00:12:27) |
|test (roxie) |1008 |1008 |0 |396 sec  (00:06:36) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  771 sec (00:12:51) |  126 sec (00:02:06) |   23 sec (00:00:23) |   19 sec (00:00:19) | 1823 sec (00:30:23) |   19 sec (00:00:19) | 2790 sec (00:46:30) |

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 718e763fc11752ae2eebaec00a647fa7387b8595
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |76 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |9 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |30 sec  (00:00:30) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |929 |929 |0 |646 sec  (00:10:46) |
|test (thor) |842 |842 |0 |1025 sec  (00:17:05) |
|test (roxie) |1008 |1007 |1 |1180 sec  (00:19:40) |


[Errors:](https://gist.github.com/ce58218ed7a002d794e21d0723eed56f)
- test
  - roxie
    - 661\. [schedule2](https://gist.github.com/HPCCSmoketest/ce58218ed7a002d794e21d0723eed56f#file-schedule2-w20200828-213258-1-log), 

HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) | 1081 sec (00:18:01) |  127 sec (00:02:07) |   23 sec (00:00:23) |   17 sec (00:00:17) | 3211 sec (00:53:31) |   17 sec (00:00:17) | 4485 sec (01:14:45) |

AttilaVamos commented:
I am sure the schedule2 failure isn't related to this PR.

afishbeck commented:
@richardkchapman @ghalliday

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: New commit.
Commit ID: 0039b497c595cecdf34b7da4ab989c7049789258
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

afishbeck commented:
Updated caching logic to check all viable caching locations before going disk or vaults.

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: 0039b497c595cecdf34b7da4ab989c7049789258
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-251.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-251.ca-central-1.compute.internal
Sha: 0039b497c595cecdf34b7da4ab989c7049789258
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |72 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |30 sec  (00:00:30) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |929 |929 |0 |331 sec  (00:05:31) |
|test (thor) |842 |842 |0 |734 sec  (00:12:14) |
|test (roxie) |1008 |1008 |0 |408 sec  (00:06:48) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  765 sec (00:12:45) |  120 sec (00:02:00) |   23 sec (00:00:23) |   19 sec (00:00:19) | 1825 sec (00:30:25) |   19 sec (00:00:19) | 2780 sec (00:46:20) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 0039b497c595cecdf34b7da4ab989c7049789258
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |74 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |929 |929 |0 |648 sec  (00:10:48) |
|test (thor) |842 |842 |0 |1023 sec  (00:17:03) |
|test (roxie) |1008 |1008 |0 |760 sec  (00:12:40) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) | 1066 sec (00:17:46) |  127 sec (00:02:07) |   23 sec (00:00:23) |   17 sec (00:00:17) | 2784 sec (00:46:24) |   18 sec (00:00:18) | 4044 sec (01:07:24) |

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3030,19 +3051,117 @@ int getEnum(const char *v, const EnumMapping *map, int defval)
+        https = false;
+    else
+        throw MakeStringException(-1, "Invalid URL, no protocol specified %s", url);
+    if (https)
```
trivial: curious why this code is separated from the code above.

afishbeck replied:
Side effect of how the code evolved, it would be better moved up.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3030,19 +3051,117 @@ int getEnum(const char *v, const EnumMapping *map, int defval)
+        throw MakeStringException(-1, "Invalid URL %s", url);
+    if (0 == strncasecmp(url, "HTTPS://", 8))
+        https = true;
+    else if (0 == strncasecmp(url, "HTTP://", 7))
```
question: are there likely to be other protocols?

afishbeck replied:
In terms of vaults I would think most are HTTPS/HTTP accessible.  In terms of general future thinking I could see us supporting protocol buffers / gRPC in the future.  I could remove the https flag for now, since I don't use it.  Or change is to an ssl flag.  I was trying to make functions that might be useful elsewhere.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+            if (created && (created < timeoutThreshold))
+            {
+                secretCache->removeProp(name);
+                puts("\nremoved from local cache\n");
```
minor: Debug tracing that should probably be removed/cleaned up.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+        if (secret)
+        {
+            unsigned created = (unsigned) secret->getPropInt("@created");
+            if (created && (created < timeoutThreshold))
```
nit picking: This could cause wrapping issues if created is a very large number - it could never time out.  You need to check if now - created > getSecretTimeout().

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+    {
+        CriticalBlock block(secretCacheCS);
+        secretCache->setPropTree(name, LINK(secret));
+        puts("\nadded to local cache\n");
```
more debug tracing.  (Should be moved out of the cs if kept).

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+}
+
+CriticalSection secretCacheCS;
+static Owned<IPropertyTree> secretCache = createPTree();
```
need to be careful about construction ordering.  I'm not sure how much of a problem it is these days, but safer to use a MODULE_INIT to initialize because jptree has some interesting initialization.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+    return vaultManager;
+}
+
+CriticalSection secretCacheCS;
```
minor: add static

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+            continue;
+        StringBuffer value;
+        value.loadFile(&dir->query());
+        tree->setProp(name, value.str());
```
What implications does this have for binary secrets?  Is it ok to assume they will never be used?

afishbeck replied:
I'll do a little research.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+
+extern jlib_decl IPropertyTree *getSecret(const char *category, const char * name)
+{
+    //check for any chached first
```
typo: cached

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+    StringAttr token;
+
+public:
+    CVault(const char *name_, IPropertyTree *vault) : name(name_)
```
trivial: code elsewhere uses a prefix of _ for clashing parameter names.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+            httplib::Client cli(schemeHostPort.str());
+            if (username.length() && password.length())
+                cli.set_basic_auth(username, password);
+            if (httplib::Result res = cli.Post("/v1/auth/kubernetes/login", json, "application/json"))
```
minor: code generally assigns first and then checks.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+            if (!envelope)
+                return false;
+            unsigned created = (unsigned) envelope->getPropInt("@created");
+            if (created && (created < timeoutThreshold))
```
minor: similar problem with wrapping as the local cache

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+            if (created && (created < timeoutThreshold))
+            {
+                tree->removeTree(envelope);
+                puts("\nremoved from vault cache\n");
```
similar comments about tracing with puts

afishbeck replied:
Forgot to scrub these personal debug statements out.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+        return false;
+    }
+};
+class CVaultSet
```
minor: newline between classes.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
-    //MORE: cache the secret for up to secretTimeoutMs
+    bool getCachedSecret(CVaultKind &kind, StringBuffer &content, const char *secret, const char *version)
+    {
+        std::map<std::string, std::unique_ptr<CVault>>::iterator it = vaults.begin();
```
possibly simpler to use "auto" instead of the explicit type (and elsewhere)

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
 }
 
-extern jlib_decl StringBuffer & getSecret(StringBuffer & result, const char * name, const char * key)
+enum class CVaultKind { kv_v1, kv_v2 };
```
is it worth splitting this code into a jsecret.hpp/cpp?  (I don't have a strong opinion either way.)

afishbeck replied:
Yes I'll split it out.  Wanted an initial sanity check on the approach first.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -3030,19 +3051,117 @@ int getEnum(const char *v, const EnumMapping *map, int defval)
 static StringBuffer secretDirectory;
 static CriticalSection secretCS;
-static unsigned secretTimeoutMs = UINT_MAX;
+static unsigned secretTimeoutMs = 60 * 60 * 1000;
```
Is this typical for k8s?  (I never researched what the standard default was.)

afishbeck replied:
I'll see if there's some consensus somewhere.

ghalliday reviewed (commented):
@afishbeck the changes look good to me.  A few comments, but none of them very serious.

afishbeck commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -313,6 +313,22 @@ Add Secret volume for a component
+  {{ $category }}:
+  {{- range $vaultid, $vault := . }}
+    {{ $vaultid }}:
+      type: {{ $vault.type }}
```
Yes, good catch I changed from type to kind but missed it here.  kv-v2 happens to be the default.

afishbeck commented on helm/examples/secrets/README.md:
````diff
@@ -0,0 +1,137 @@
+Enabling kubernetes auth will allow k8s nodes to access the vault via their kubernetes.io access tokens.
+
+```bash
+vault auth enable kubernetes
````
I've moved the login up higher and now enable kubernetes should work from outside the vault-0 pod.

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: ab0ead07e789f38fbd77c6eae49304a90f8ea094
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-120.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-120.ca-central-1.compute.internal
Sha: ab0ead07e789f38fbd77c6eae49304a90f8ea094
Conflicting files, should skip build and test.

[Build error:](https://gist.github.com/3733060ba11b44fc6a3751c9dac4cd53)
Build log [Build log](https://gist.github.com/HPCCSmoketest/3733060ba11b44fc6a3751c9dac4cd53#file-result-20-09-04-13-07-07-log)

afishbeck commented:
Updated to support binary secrets, moved secret code to jsecrets.cpp, cleaned up based on review comments.
@gfortil updated README.md based on your comments.  Among other things I now assume user is running from HPCC-Platform/helm.
@stuartort if you want to test when you get to the hpcc helm install line add " --set global.image.root=afishbeck" as follows:

helm install myhpcc hpcc/  --set global.image.root=afishbeck --set global.image.version=latest -f examples/secrets/values-secrets.yaml

afishbeck commented:
@AttilaVamos not sure what "Conflicting files, should skip build and test" means.  I'll try rebasing.

afishbeck commented:
rebased

afishbeck commented:
Also changed the vault yaml format a bit.
@ghalliday please review

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: 51e88553e26ff3d0aee6eb829a4464d0d7f9d145
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-131.ca-central-1.compute.internal

afishbeck commented:
rebased again to resolve conflict with my other PR which was just merged.

afishbeck commented:
@ghalliday

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-131.ca-central-1.compute.internal
Sha: 51e88553e26ff3d0aee6eb829a4464d0d7f9d145
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |76 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |30 sec  (00:00:30) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |25 sec  (00:00:25) |
|test (hthor) |930 |930 |0 |343 sec  (00:05:43) |
|test (thor) |842 |842 |0 |749 sec  (00:12:29) |
|test (roxie) |1009 |1009 |0 |422 sec  (00:07:02) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  772 sec (00:12:52) |  121 sec (00:02:01) |   23 sec (00:00:23) |   18 sec (00:00:18) | 1874 sec (00:31:14) |   19 sec (00:00:19) | 2836 sec (00:47:16) |

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: a4d572450f76bd9aa040ca6cec905745284cb29e
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-94.ca-central-1.compute.internal

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: New commit.
Commit ID: ce85dc66c91006814efc8cd0db5c529a4fccf168
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-94.ca-central-1.compute.internal
Sha: a4d572450f76bd9aa040ca6cec905745284cb29e
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |74 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |42 sec  (00:00:42) |
|setup (roxie) |11 |11 |0 |19 sec  (00:00:19) |
|test (hthor) |930 |930 |0 |345 sec  (00:05:45) |
|test (thor) |842 |842 |0 |754 sec  (00:12:34) |
|test (roxie) |1009 |1009 |0 |398 sec  (00:06:38) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  782 sec (00:13:02) |  124 sec (00:02:04) |   23 sec (00:00:23) |   19 sec (00:00:19) | 1845 sec (00:30:45) |   18 sec (00:00:18) | 2821 sec (00:47:01) |

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: ce85dc66c91006814efc8cd0db5c529a4fccf168
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-165.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: ce85dc66c91006814efc8cd0db5c529a4fccf168
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |76 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |26 sec  (00:00:26) |
|test (hthor) |930 |930 |0 |661 sec  (00:11:01) |
|test (thor) |842 |842 |0 |1031 sec  (00:17:11) |
|test (roxie) |1009 |1009 |0 |745 sec  (00:12:25) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) | 1058 sec (00:17:38) |  128 sec (00:02:08) |   23 sec (00:00:23) |   17 sec (00:00:17) | 2801 sec (00:46:41) |   18 sec (00:00:18) | 4054 sec (01:07:34) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-165.ca-central-1.compute.internal
Sha: ce85dc66c91006814efc8cd0db5c529a4fccf168
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |75 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |930 |930 |0 |339 sec  (00:05:39) |
|test (thor) |842 |842 |0 |747 sec  (00:12:27) |
|test (roxie) |1009 |1009 |0 |423 sec  (00:07:03) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  779 sec (00:12:59) |  121 sec (00:02:01) |   23 sec (00:00:23) |   17 sec (00:00:17) | 1860 sec (00:31:00) |   19 sec (00:00:19) | 2828 sec (00:47:08) |

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: 008d8701e84ad14536c9e1c87d8fa514598fffa4
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-221.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-221.ca-central-1.compute.internal
Sha: 008d8701e84ad14536c9e1c87d8fa514598fffa4
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |73 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |42 sec  (00:00:42) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |930 |930 |0 |336 sec  (00:05:36) |
|test (thor) |842 |842 |0 |750 sec  (00:12:30) |
|test (roxie) |1009 |1009 |0 |396 sec  (00:06:36) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  762 sec (00:12:42) |  120 sec (00:02:00) |   23 sec (00:00:23) |   18 sec (00:00:18) | 1831 sec (00:30:31) |   18 sec (00:00:18) | 2781 sec (00:46:21) |

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: New commit.
Commit ID: 008d8701e84ad14536c9e1c87d8fa514598fffa4
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 008d8701e84ad14536c9e1c87d8fa514598fffa4
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |76 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |25 sec  (00:00:25) |
|test (hthor) |930 |930 |0 |641 sec  (00:10:41) |
|test (thor) |842 |842 |0 |1021 sec  (00:17:01) |
|test (roxie) |1009 |1007 |2 |1480 sec  (00:24:40) |


[Errors:](https://gist.github.com/6abea9d92880b42d4ecef944f11c5465)
- test
  - roxie
    - 662\. [schedule2](https://gist.github.com/HPCCSmoketest/6abea9d92880b42d4ecef944f11c5465#file-schedule2-w20200904-212258-5-log), 992\. [workflow_contingency_8 (parallel=true)](https://gist.github.com/HPCCSmoketest/6abea9d92880b42d4ecef944f11c5465#file-workflow_contingency_8-parallel-true-w20200904-212800-log), 

HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) | 1086 sec (00:18:06) |  127 sec (00:02:07) |   24 sec (00:00:24) |   18 sec (00:00:18) | 3527 sec (00:58:47) |   17 sec (00:00:17) | 4808 sec (01:20:08) |

afishbeck commented:
@AttilaVamos I doubt those roxie timeouts had much to do with this change, can we re-run the test?

AttilaVamos commented:
@afishbeck The schedule2 is definitely not related to this PR. The workflow test is new, I should check it. However I think it isn't related, as well based on that test has been successful on in the another  Smoketest environment (Host: ip-10-20-0-221.ca-central-1.compute.internal).

ghalliday commented:
Added HPCC-24707 to cover the last issue - not likely to be related to tony's change, but worth investigating.

ghalliday commented:
@ghalliday - to make sure I review it next

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: 22f370062a04c46afefdeb4316ff323167d499b1
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-115.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-115.ca-central-1.compute.internal
Sha: 22f370062a04c46afefdeb4316ff323167d499b1
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |72 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |17 sec  (00:00:17) |
|test (hthor) |930 |930 |0 |339 sec  (00:05:39) |
|test (thor) |842 |842 |0 |729 sec  (00:12:09) |
|test (roxie) |1009 |1009 |0 |414 sec  (00:06:54) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  766 sec (00:12:46) |  121 sec (00:02:01) |   23 sec (00:00:23) |   19 sec (00:00:19) | 1827 sec (00:30:27) |   18 sec (00:00:18) | 2783 sec (00:46:23) |

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: New commit.
Commit ID: 22f370062a04c46afefdeb4316ff323167d499b1
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 22f370062a04c46afefdeb4316ff323167d499b1
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |73 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |9 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |930 |930 |0 |648 sec  (00:10:48) |
|test (thor) |842 |842 |0 |1013 sec  (00:16:53) |
|test (roxie) |1009 |1009 |0 |755 sec  (00:12:35) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) | 1078 sec (00:17:58) |  126 sec (00:02:06) |   23 sec (00:00:23) |   19 sec (00:00:19) | 2769 sec (00:46:09) |   19 sec (00:00:19) | 4054 sec (01:07:34) |

ghalliday commented on system/jlib/jsecrets.cpp:
```diff
@@ -0,0 +1,702 @@
+        if (token.length())
+            return;
+        StringBuffer login_token;
+        login_token.loadFile("/var/run/secrets/kubernetes.io/serviceaccount/token");
```
Does this require the pods to have any access rights?  Some recent PRs of Richard's have tied down both the access rights and the network access.  Does that introduce problems here?

afishbeck replied:
Based on a quick comment from @richardkchapman I think the line in the README where I do:

vault write auth/kubernetes/role/hpcc-vault-access \
        bound_service_account_names=hpcc-default \
        bound_service_account_namespaces=default \
        policies=hpcc-kv-ro \
        ttl=24h
may be enough to grant access.
It's more of an operational thing on the vault side granting access to our default account, or whatever the user may change the account to.  
I also support passing in the authentication token as a k8s secret.  Hopefully that will cover some other authentication schemes the user may set up.  The user can take whatever steps necessary to generate that.
Vault supports many authentication schemes and we can add more over time based on what scenarios users need, or we want to support in general.

ghalliday reviewed (commented):
@afishbeck this looks like it is very close to being ready.  A few questions/comments.

ghalliday commented on helm/hpcc/values.schema.json:
```diff
@@ -55,6 +55,31 @@
+        "ecl": {
+          "$ref": "#/definitions/vaultCategory"
+        },
+        "ecl-user": {
```
We should consider publishing the k8s secrets in subdirectories corresponding to the category - I think that would avoid the issue.

afishbeck replied:
The categories allow the various types of secrets to be organized and secured differently. To prevent leaking and usage for unintended purposes.  I haven't made use of "ecl-user" yet, but that is going to be the least secure category, ecl code will have direct access to the contents of the secret.  For example, if I need a key for use with some plugin or to pass in some non standard way to a 3rd party.  Because the ecl can see the contents it could return it, write it, or send it anywhere.  I don't want gateway or storage plane credentials to be accessible in that way.

When it comes to the organization of the actual vault it could be a simple as:
http://${env.VAULT_SERVICE_HOST}:${env.VAULT_SERVICE_PORT}/v1/secret/data/storage/${secret}
http://${env.VAULT_SERVICE_HOST}:${env.VAULT_SERVICE_PORT}/v1/secret/data/ecl/${secret}
http://${env.VAULT_SERVICE_HOST}:${env.VAULT_SERVICE_PORT}/v1/secret/data/ecl-user/${secret}

But those different sections can have different access rights, and allow the secrets to be organized based on how and where they are used.

Over time we may want to segment the vault access by components.  So certain roxies can call some gateways but hthor can't for example.  But I'd rather get a feel for how these things are used in the wild a bit first.  Our users can give us some feedback on how they need to further restrict.

I definitely agree on the idea of the k8s secret subdirectories.  That might mean I no longer have to mangle the names of the HTTP-CONNECT secrets.  That was for example to prevent them being used in the way the ecl-user category will be in the future.

afishbeck commented on common/thorhelper/thorsoapcall.cpp:
```diff
@@ -801,10 +805,13 @@ class CWSCHelper : implements IWSCHelper, public CInterface
     unsigned done;
     Owned<IPropertyTree> xpathHints;
     Linked<ClientCertificate> clientCert;
+    bool customClientCert = false;
```
Yes, if I need to make any other changes I will include this change as well.

afishbeck commented:
@ghalliday updated

1. Secret categories are based on usage, not just where they should go
2. Added new system category for HPCC system level secrets.
3. k8s secret path includes category.
4. Fixed vault configMap entries for THOR
5. README now demonstrates enabling all account names.
6. Minor improvements.

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: Forced to re-test.
Commit ID: 7689bf4acddafb1b978f2787518bba44c3b940b2
Estimated completion time is ~0.8 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-50.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-50.ca-central-1.compute.internal
Sha: 7689bf4acddafb1b978f2787518bba44c3b940b2
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |75 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |19 sec  (00:00:19) |
|test (hthor) |930 |930 |0 |337 sec  (00:05:37) |
|test (thor) |842 |842 |0 |739 sec  (00:12:19) |
|test (roxie) |1009 |1009 |0 |396 sec  (00:06:36) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  773 sec (00:12:53) |  120 sec (00:02:00) |   23 sec (00:00:23) |   19 sec (00:00:19) | 1824 sec (00:30:24) |   18 sec (00:00:18) | 2786 sec (00:46:26) |

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: New commit.
Commit ID: 7689bf4acddafb1b978f2787518bba44c3b940b2
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 7689bf4acddafb1b978f2787518bba44c3b940b2
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |74 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |31 sec  (00:00:31) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |930 |930 |0 |644 sec  (00:10:44) |
|test (thor) |842 |842 |0 |1012 sec  (00:16:52) |
|test (roxie) |1009 |1009 |0 |750 sec  (00:12:30) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) | 1065 sec (00:17:45) |  127 sec (00:02:07) |   23 sec (00:00:23) |   17 sec (00:00:17) | 2785 sec (00:46:25) |   18 sec (00:00:18) | 4044 sec (01:07:24) |

ghalliday commented:
@afishbeck I like the changes - please can you squash and I will scan it one more time before merging.

afishbeck commented:
@ghalliday squashed changes

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-181.ca-central-1.compute.internal
Sha: 6b9063a58436fc7731a3788919d9e6ef8e4a7153
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |75 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |16 sec  (00:00:16) |
|test (hthor) |930 |930 |0 |330 sec  (00:05:30) |
|test (thor) |842 |842 |0 |742 sec  (00:12:22) |
|test (roxie) |1009 |1009 |0 |424 sec  (00:07:04) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  778 sec (00:12:58) |  124 sec (00:02:04) |   23 sec (00:00:23) |   18 sec (00:00:18) | 1843 sec (00:30:43) |   18 sec (00:00:18) | 2814 sec (00:46:54) |

HPCCSmoketest commented:
Process of PR-14122, label: secrets_and_vaults starts now.
The reason of this test is: New commit.
Commit ID: 6b9063a58436fc7731a3788919d9e6ef8e4a7153
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 6b9063a58436fc7731a3788919d9e6ef8e4a7153
Build: success
Milestone:Install hpccsystems-platform-community_7.11.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |144 |144 |0 |0 |0 |74 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |930 |930 |0 |664 sec  (00:11:04) |
|test (thor) |842 |842 |0 |1036 sec  (00:17:16) |
|test (roxie) |1009 |1009 |0 |742 sec  (00:12:22) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) | 1097 sec (00:18:17) |  126 sec (00:02:06) |   23 sec (00:00:23) |   18 sec (00:00:18) | 2797 sec (00:46:37) |   18 sec (00:00:18) | 4088 sec (01:08:08) |
